File: rexup/src/globalTexts.ts
import type { FileOrDirectory } from "./components/types";

// Is used to enhance the displayed quality of paths in the frontend.
// It needs to be rendered as HTML to take effect.
function wrapInMonospaceBox(value: string) {
	return `<span class="bg-gray-800 rounded-md px-1 opacity-75">${value}</span>`;
}

export const globalTexts = {
	sidebar: {
		heading: "Backups",
		addBackup: "Create backup",
		addBackupPopup: {
			heading: "Create backup",
			description:
				"Create a new backup, that is added to the list on the sidebar. Please enter a name to proceed.",
			placeholder: "A fantastic name",
			confirm: "Create",
		},
		settingsPopup: {
			heading: "Settings",
			toggleDisplayHistoryLabel: "Show backup execution history",
			toggleDisplayHistoryDescription:
				"Enable or disable the appearance of the execution history.",
			deleteAllDataLabel: "Delete all your data",
			deleteAllDataDescription:
				"This includes your config file, the recorded history of backup executions and the structure of backups but NOT the actual backups that were created.",
			deleteAllDataButton: "Delete all data",
		},
	},
	overview: {
		pathSelectorPopup: {
			pathElementsHeading: "Path",
			bookmarks: {
				heading: "Bookmarks:",
				home: "Home",
				desktop: "Desktop",
				downloads: "Downloads",
				documents: "Desktop",
			},
			results: {
				heading: "Nested directories in the current directory:",
				noResults: {
					info: "There aren't any entries in this directory",
					infoEmoji: ":/",
				},
			},
			selectPathConfirm: "Select path",
		},
		noBackupSelected: {
			heading: "There's no backup selected",
			headingEmoji: ":/",
			description:
				"Click on an existing backup or create one to edit it in here.",
		},
		backupOverview: {
			heading: (name: string) => {
				return `Overview of backup "${name}"`;
			},
			executeBackup: "Execute backup",
			switchToLogTabHint:
				"After you clicked the button above, switch to the Logs tab (if you haven't already) to stay informed about the progress of the backup execution.",
			tabSwitcher: {
				entries: "Entries",
				logs: "Logs",
				configuration: "Configuration",
			},
		},
		listIcon: "\u{2022}", // U+2022 or "•"
		configurationTab: {
			heading: "Configuration",
			nameSection: {
				label: "Backup name",
				description: "Adjust the name display in the sidebar.",
				placeholder: "A creative name",
			},
			isZippedSection: {
				label: "Output variant",
				description:
					"Your backup can either be a simple folder containing nested folders or files or it can be a zip file. Zip files have the advantage of being smaller and easier to transfer between devices. On the other hand, zip files need to be extracted if you want to work with the contained data.",
				directoryOption: "Folder",
				zipFileOption: "Zip file",
			},
			locationSection: {
				label: "Backup location",
				description:
					'After your backup has been executed, its placed somewhere. This location can be configured here. To change it, simply click the "Edit" button below and choose a directory the backup is placed into.',
				inputPlaceholder: "Your desktop",
				pathSelectorHeading: "Select backup location",
				editInput: "Edit",
				resetInput: "Reset",
				accessGranted:
					"Everything is fine! You have access to this location on your file system.",
				accessProhibited:
					"Something is not working! You don't seem to have access to this location on your file system. Please choose a different directory for the backup to work properly.",
			},
			destructiveSection: {
				label: "Delete backup",
				description:
					'To delete this backup, click the "Delete this backup" button below. This deletes the entire backup structure but NOT the actual backups themselves.',
				delete: "Delete this backup",
				reallyDelete: "REALLY delete this backup?",
			},
		},
		logsTab: {
			heading: "Logs",
			emptyList: "Currently, this list is empty.",
			executionHistory: {
				label: "Execution history",
				description:
					"These logs keep track of the most recent backup executions.",
			},
			executionLogs: {
				label: "Execution logs",
				description:
					"The list below displays details of the last backup execution like errors and ignored files.",
				icons: {
					finished: "\u{1F3C1}", // U+1F3C1 or "🏁"
					information: "\u{2139}", // U+2139 or "ℹ️"
					error: "\u{274C}", // U+274C or "❌"
					success: "\u{2705}", // U+2705 or "✅"
					ignore: "\u{1F6AB}", // U+1F6AB or "🚫"
				},
				successfullCommandLog: (command: string, toPath: string) => {
					return `Executed the command ${wrapInMonospaceBox(command)} at ${wrapInMonospaceBox(toPath)} successfully.`;
				},
				successfullLocalFileSystemLog: (
					variant: FileOrDirectory,
					fromPath: string,
					toPath: string,
				) => {
					return `Copied the ${variant.toLocaleLowerCase()} from ${wrapInMonospaceBox(fromPath)} to ${wrapInMonospaceBox(toPath)} successfully.`;
				},
				ignoreFileLog: (fromPath: string, toPath: string, reason: string) => {
					return `Ignored copying the file from ${wrapInMonospaceBox(fromPath)} to ${wrapInMonospaceBox(toPath)} because the file ${reason}.`;
				},
				ignoreReasons: {
					tooLarge: "is too large",
					wrongName: "has the wrong name",
					wrongPathElements: "has the wrong path elements",
				},
			},
		},
		entriesTab: {
			heading: "Entries",
			createBackupEntry: "Create backup entry",
			pathSelectorHeading: "Select origin location",
			backupEntryPreview: {
				entryEnabled: "Enabled",
				entryDisabled: "Disabled",
				edit: "Edit",
				localFileSystemPreviewText: (variant: FileOrDirectory | null) => {
					return `This backup entry currently copies a ${variant === null ? "unknown thing" : variant.toLocaleLowerCase()} at following location:`;
				},
				commandPreviewText: (length: number) => {
					return `This backup entry currently executes ${length === 1 ? "one command" : `${length} commands`}.`;
				},
			},
			addBackupEntryPopup: {
				heading: "Create backup entry",
				description:
					"Create a new backup entry that is added to the list on the Overview tab. Please enter a name to proceed.",
				placeholder: "A fantastic name",
				confirm: "Create",
			},
			editBackupEntryPopup: {
				heading: "Edit backup entry",
				navbarIcons: {
					overview: "Overview",
					origin: "Origin",
					target: "Target",
					filters: "Filters",
					destructive: "Destructive",
				},
				assistanceReference:
					"For further assistance, switch to the Overview tab, which explains the copy process in detail.",
				destructiveTab: {
					heading: "Delete backup entry",
					description:
						'To delete this backup entry, click the "Delete this backup entry" button below. This deletes the entire backup entry structure but NOT the actual backups themselves.',
					delete: "Delete this backup entry",
					reallyDelete: "REALLY delete this backup entry?",
				},
				overviewTab: {
					nameLabel: "Name:",
					explanationLabel: "Explanation:",
					localFileSystemMode: {
						description:
							'Currently, the "LocalFileSystem" mode is enabled. This means that this backup entry copies files and directories from your device into the backup.',
						noOrigin:
							"Because there's no file or directory selected, a more detailed explanation can't be generated.",
						detailedExplanation: (
							variant: FileOrDirectory,
							origin: string,
							target: string,
						) => {
							return `In this case, the ${variant.toLocaleLowerCase()} located at ${wrapInMonospaceBox(origin)} on your device is copied into ${target === "" || target === "/" ? "the root directory" : `${wrapInMonospaceBox(target)}`} of the backup.`;
						},
						renameExplanation: (renameTo: string) => {
							return `Additionally, it will be renamed to ${wrapInMonospaceBox(renameTo)} after it has been placed in this directory.`;
						},
						filterExplanation: (
							countOfActiveFilters: number,
							filtersInPretty: string,
						) => {
							return `Furthermore, ${countOfActiveFilters === 1 ? "one filter" : `${countOfActiveFilters} filters`} (${filtersInPretty}) ${countOfActiveFilters > 1 ? "are" : "is"} active which may ${countOfActiveFilters > 1 ? "prevent" : "prevents"} files from being copied. Switch to the Filters tab to see the precise settings.`;
						},
						filterAbbreviations: {
							maximumFileSize: "maximum size",
							fileNames: "file names",
							pathElements: "path elements",
						},
					},
					commandsMode: {
						description:
							'Currently, the "Commands" mode is enabled. This means that this backup entry executes the specified commands inside the backup at the target location.',
						noCommands:
							"Because there aren't any commands that should be executed, a more detailed explanation can't be generated.",
						detailedExplanation: (count: number, target: string) => {
							return `In this case, the following ${count === 1 ? "one command is" : `${count} commands are`} executed at "${target}":`;
						},
					},
				},
				originTab: {
					heading: "Origin:",
					description:
						"Here you can select the resource to copy. It can either be a file or directory on your local device or a list of commands that e.g. clone Git repositories.",
					mode: {
						heading: "Mode:",
						localFileSystemOption: "Local file system",
						commandsOption: "Commands",
					},
					actualInput: {
						heading: "Value:",
						localFileSystem: {
							placeholder: "Unset",
							edit: "Edit",
							reset: "Reset",
						},
						commandListPlaceholder: "Command to execute",
					},
				},
				targetTab: {
					placeholder: "Unset",
					reset: "Reset",
					target: {
						heading: "Target:",
						description:
							'Here you can select the location of the copied resource relative to the directory or zip file that has been created. If the input field is empty, the resource is placed in the root directory (= "/").',
					},
					rename: {
						heading: "Rename to:",
						description:
							"Here you can pick the new name of the copied resource. If the input field is empty, the resource isn't renamed and keeps its original name. Note that the rename functionality only works on files and directories that are stored on your local file system.",
					},
				},
				filtersTab: {
					heading: "Filters:",
					description:
						"Filters only work if the origin points to a file or directory in the local file system.",
					maximumFileSize: {
						heading: "Maximum file size:",
						description:
							"This filter describes the maximum file size in MB a file is allowed to have. If a file has a greater size, it's ignored. This filter also applies to directories where every file is checked seperately.",
						placeholder: "Maximum size",
					},
					filterMode: {
						heading: "File names/Path elements mode:",
						description:
							'Choose whether files matching the following two filters should or shouldn\'t be included in the backup. This control has no effect on the "Maximum file size" filter.',
						includeOption: "Include",
						excludeOption: "Exclude",
					},
					fileNames: {
						heading: "File names:",
						description:
							"This filter lists the names a file is allowed or disallowed to have to be copied. If the input field is empty, the filter is inactive.",
						placeholder: "File name",
					},
					pathElements: {
						heading: "Path elements:",
						description:
							"This filter lists the elements a file is allowed or disallowed to have in its path to be copied. If the input field is empty, the filter is inactive.",
						placeholder: "Path element",
					},
				},
			},
		},
	},
};

File: rexup/src-tauri/src/global_texts.rs
//! Contains texts that are displayed to the user in the application. Therefore, it is collected in one place to gain an overview.
//! Currently, there are only texts for `BackupExecutionLog::ErrorCopying` and `BackupExecutionLog::Information` placed in this module.

use std::{ io::Error, path::{ Path, PathBuf } };

pub fn backup_parent_not_created(path: &str) -> String {
	// Replaces and empty path with user friendly text which is better than just displaying an empty path if the backup location is not modified.
	let real_path: String;

	if path.is_empty() {
		real_path = "your desktop".to_string();
	} else {
		real_path = format!("\"{path:?}\"");
	}

	format!(
		"Can't create the backup parent directory at \"{real_path}\" so the backup can't be executed."
	)
}

pub const ZIP_WRITER_NOT_CREATED: &str =
	"Can't create the zip writer so the backup can't be executed.";

pub fn skipped_backup_entry(name: &str) -> String {
	format!("Skipped execution of entry \"{name}\" because it's disabled.")
}

pub fn started_backup_entry(name: &str) -> String {
	format!("Started execution of entry \"{name}\".")
}

pub fn finished_backup_entry(name: &str) -> String {
	format!("Finished execution of entry \"{name}\".")
}

pub const TEMP_DIR_NOT_CREATED: &str =
	"Can't create a temporary directory to store the results of the executed commands before copying them into the zip file. Therefore, the execution of this backup entry has been cancelled.";

pub const TEMP_DIR_NOT_DELETED: &str =
	"Can't delete the temporary desktop on the user's desktop after zipping all the files.";

pub fn subdirectories_for_commands_in_backup_parent_not_created(
	relative_target: &Path,
	commands: &Vec<String>
) -> String {
	format!(
		"The subdirectories at \"{relative_target:?}\" inside the backup parent directory can't be created. Therefore, the command(s) \"{commands:?}\" can't be executed."
	)
}

pub fn subdirectories_for_file_in_backup_parent_not_created(
	relative_target: &Path,
	origin: &Path
) -> String {
	format!(
		"The subdirectories at \"{relative_target:?}\" inside the backup parent directory can't be created. Therefore, the file at \"{origin:?}\" can't be copied."
	)
}

pub fn commands_in_backup_parent_not_executed(
	corrected_relative_target: &Path,
	command: &str
) -> String {
	format!(
		"The command \"{command}\" at \"{corrected_relative_target:?}\" inside the backup parent directory couldn't be executed. Therefore, its skipped."
	)
}

pub fn origin_name_not_found(origin: &Path) -> String {
	format!(
		"Can't get the origin's file or directory name of \"{origin:?}\". Therefore, copying this file or directory is not possible."
	)
}

pub fn origin_not_found(origin: &Path) -> String {
	format!("The location at \"{origin:?}\" doesn't exist and therefore can't be copied.")
}

pub fn prefix_not_stripped(target: &str) -> String {
	format!(
		"Can't strip the leading \"/\" of \"{target}\". Therefore, copying this backup entry is not possible."
	)
}

pub fn file_not_copied(origin: &Path, target: PathBuf) -> String {
	format!(
		"The file at \"{origin:?}\" can't be be written to \"{target:?}\". Therefore, this file can't be copied."
	)
}

pub fn file_not_read(path: &Path) -> String {
	format!("The file at \"{path:?}\" can't be be read and therefore can't be copied.")
}

pub fn directory_not_read(path: &Path) -> String {
	format!(
		"The directory at \"{path:?}\" can't be read. Therefore, copying this directory is not possible."
	)
}

pub fn entry_not_read(error: Error, path: &Path) -> String {
	format!(
		"An entry \"{error:?}\" in \"{path:?}\" can't be read. Therefore, reading and possibly copying this entry is not possible."
	)
}

pub fn file_in_zip_writer_not_started(
	relative_target_and_file_name: &PathBuf,
	origin: &Path
) -> String {
	format!(
		"The zip writer can't start a new file at \"{relative_target_and_file_name:?}\". Therefore, the file at \"{origin:?}\" can't be copied."
	)
}

pub fn file_in_zip_writer_not_written_to(
	relative_target_and_file_name: &PathBuf,
	origin: &Path
) -> String {
	format!(
		"The zip writer can't write to the file at \"{relative_target_and_file_name:?}\" inside the zip file. Therefore, this file at \"{origin:?}\" can't be copied."
	)
}
